### PR TITLE
Issue #5077: Distinguish layout-created/module-provided/missing-callback paths on the layout listing page

### LINKED ISSUE #5077: Added: Distinguish layout-created/module-provided/missing-callback paths on the layout listing page.
This is a follow-on to https://github.com/backdrop/backdrop-issues/issues/4986, which noted that for pages that have both a module-defined page and a layout with the same path, the page callback that actually gets called depends on the chronological order of the module-enabling and layout-creation that happened in the past. In that issue, I question whether we should have that past-history dependence. In this issue, I accept that dependence, but propose some changes that will make it easier for users to figure out what is going on.

Specifically, I propose to add some formatting to the system paths shown in the first column on the layout listing page at admin/structure/layouts:

1. If the page will be rendered by a module, use the current formatting (bold) for the path.
2. If the page will be rendered by a layout, use bold+italic.
3. If the page won't be rendered at all because there is no menu router item for the page, use bold+red.

Currently, there is no way to tell which of these three situations applies, short of attempting to go to the page and seeing what renders (or 404s, for case 3), which can be more puzzling than illuminating. If you apply the PR below, it's a lot easier to distinguish the different scenarios in https://github.com/backdrop/backdrop-issues/issues/4986. With regard to the 5 scenarios in https://github.com/backdrop/backdrop-issues/issues/4986, you should see:

* Scenario 1: bold (module provides the page)
* Scenario 2: bold+italic (layout provides the page, hiding module page)
* Scenario 3: bold+italic (layout provides the page)
* Scenario 4: bold+red (no menu router item for page, so 404)
* Scenario 5: bold+italic (layout provides the page)

I might also suggest that on the configure layout page, after the user has typed in a path, the AJAX call displays whether that page will be layout-provided or is an existing module-provided page, but that behavior isn't part of the PR below.

-- 

PR by @jenlampton, @bugfolder https://github.com/backdrop/backdrop/pull/3652 separates layouts into two tables
PR by @bugfolder https://github.com/backdrop/backdrop/pull/3864 labels each layout in a single table

### PR DESCRIPTION
**Pull request by bugfolder**
Fixes https://github.com/backdrop/backdrop-issues/issues/5077.

This is a rebase of https://github.com/backdrop/backdrop/pull/3864.

### CODE REVIEW

File: core/modules/layout/layout.admin.inc [hunk 1]
```diff
@@ -120,15 +178,34 @@ function layout_list_page() {
...
+    '#markup' => '<h2>' . t('Layout pages') . '</h2>' . t('These layouts will create new stand-alone pages on the site. Without the layouts, these pages would not exist.'),
+    '#weight' => 0,
+  );
+  $page['stand_alone_table'] = array(
     '#theme' => 'table',
     '#header' => $header,
-    '#rows' => $rows,
+    '#rows' => $stand_alone_rows,
+     '#attributes' => array('class' =>   array('layout-list')),
+     '#empty' => t('No layouts have been created yet.'),
```
Comment (klonos) [inline core/modules/layout/layout.admin.inc]:
```suggestion
     '#empty' => t('No layout pages have been created yet.'),
```

File: core/modules/layout/layout.admin.inc [hunk 2]
```diff
@@ -120,15 +178,34 @@ function layout_list_page() {
...
+    '#type' => 'help',
+    '#markup' => '<h2>' . t('Layout overrides') . '</h2>' . t('These layouts will override existing pages on the site. Without the layouts, these pages would use other layouts.'),
+    '#weight' => 2,
+   );
+   $page['override_table'] = array(
+    '#theme' => 'table',
+    '#header' => $header,
+    '#rows' => $override_rows,
     '#attributes' => array('class' =>   array('layout-list')),
     '#empty' => t('No layouts have been created yet.'),
```
Comment (klonos) [inline core/modules/layout/layout.admin.inc]:
```suggestion
    '#empty' => t('No layout overridess have been created yet.'),
```

File: core/modules/layout/layout.admin.inc [hunk 3]
```diff
@@ -120,15 +178,34 @@ function layout_list_page() {
...
     '#header' => $header,
-    '#rows' => $rows,
+    '#rows' => $stand_alone_rows,
+     '#attributes' => array('class' =>   array('layout-list')),
+     '#empty' => t('No layouts have been created yet.'),
+     '#weight' => 1,
+   );
+   $page['override'] = array(
+    '#type' => 'help',
+    '#markup' => '<h2>' . t('Layout overrides') . '</h2>' . t('These layouts will override existing pages on the site. Without the layouts, these pages would use other layouts.'),
```
Comment (klonos) [inline core/modules/layout/layout.admin.inc]:
```suggestion
    '#markup' => '<h2>' . t('Layout overrides') . '</h2>' . t('These layouts will override existing pages on the site. Without these layout overrides, these pages would use other layouts.'),
```

File: core/modules/layout/layout.admin.inc [hunk 4]
```diff
@@ -120,15 +178,34 @@ function layout_list_page() {
...
+    '#markup' => '<h2>' . t('Layout overrides') . '</h2>' . t('These layouts will override existing pages on the site. Without these layout overrides, these pages would use other layouts.'),
+    '#weight' => 2,
+   );
+   $page['override_table'] = array(
+    '#theme' => 'table',
+    '#header' => $header,
+    '#rows' => $override_rows,
     '#attributes' => array('class' =>   array('layout-list')),
-    '#empty' => t('No layouts have been created yet.'),
+    '#empty' => t('No layout overridess have been created yet.'),
```
Comment (klonos) [inline core/modules/layout/layout.admin.inc]: Ooops:

```suggestion
    '#empty' => t('No layout overrides have been created yet.'),
```

File: core/modules/layout/css/layout.admin.css [hunk 5]
```diff
@@ -47,6 +51,13 @@
   margin-left: 0;
 }
 
+.layout-list thead .layout-title {
+   padding-left: 71px;
+ }
+ [dir="rtl"] .layout-list thead .layout-title {
+   padding-right: 71px;
```
Comment (yorkshire-pudding) [inline core/modules/layout/css/layout.admin.css]: chrome_6xsA3AdHS3
The `padding-left` addition makes sense when the layout icons are there as it aligns the text of the heading with the Layout name.  However, when the width reduces and the icons are hidden, then it indents the heading compared to the layout name.
Please can this condition use the breakpoint, which is `768px`

File: core/modules/layout/layout.admin.inc [hunk 6]
```diff
@@ -14,13 +14,47 @@ function layout_list_page() {
...
 
   // Group layouts by path.
+  $layout_paths = array();
   $path_groups = array();
+  $path_groups_end = array();
   foreach ($layouts as $layout) {
-    $path_groups[$layout->getPath()][$layout->name] = $layout;
+    $path = $layout->getPath();
+     // Move admin layouts to the bottom of the list. Note that default layout
+     // has a NULL path, which path_is_admin() won't like in PHP 8.1.
```
Comment (yorkshire-pudding) [inline core/modules/layout/layout.admin.inc:23]: Noted that this is covered by #3958 (Issue [#5501](https://github.com/backdrop/backdrop-issues/issues/5501) )

File: core/modules/layout/layout.admin.inc [hunk 7]
```diff
@@ -82,19 +134,29 @@ function layout_list_page() {
...
     array('data' => t('Operations'), 'class' => array('layout-operations')),
   );
 
-  $help = '<p>' . t('As a page is rendered, a <em>Layout</em> is selected as follows.') . '</p>';
+  $help = '<p>' . t('There are two kinds of <em>Layouts</em>:') . '</p>';
+  $help_items[] = t('<strong>Layout pages</strong> are layouts that create new stand-alone pages on the site.  Without the layouts, these pages would not exist.');
+  $help_items[] = t('<strong>Layout overrides</strong> are layouts that override existing pages on the site. Without these layout overrides, these pages would use other layouts. Which one is used will depend on the path and visibility conditions. If no other layouts match, one of the defaults will be used.');
+  $help .= theme('item_list', array('items' => $help_items));
+  $help .= '<p>' . t('As a page is rendered, a <em>Layout</em> is selected as follows.') . '</p>';
   $help_items = array();
```
Comment (yorkshire-pudding) [inline core/modules/layout/layout.admin.inc:161]: `$help_items = array();` should be before line 156 (I think, though doesn't seem to affect operation):
```
 $help_items[] = t('<strong>Layout pages</strong> are layouts that create new stand-alone pages on the site.  Without the layouts, these pages would not exist.');
  $help_items[] = t('<strong>Layout overrides</strong> are layouts that override existing pages on the site. Without these layout overrides, these pages would use other layouts. Which one is used will depend on the path and visibility conditions. If no other layouts match, one of the defaults will be used.');
```

**Review by yorkshire-pudding — CHANGES_REQUESTED** [review]
Mostly looks good.  Couple of minor points.
The middle comment is just for mine and core committers benefit and doesn't require action

File: core/modules/layout/css/layout.admin.css (hunk 5 above)
Comment (bugfolder) [reply core/modules/layout/css/layout.admin.css]: Yes, good point. Breakpoint is added.

File: core/modules/layout/layout.admin.inc (hunk 7 above)
Comment (bugfolder) [reply core/modules/layout/layout.admin.inc:161]: Noted, added above line 156.

**Review by yorkshire-pudding — COMMENTED** [review]
Looks good.  Thanks @bugfolder

**Comment by quicksketch** [comment]
Thanks for the extensive reviews @yorkshire-pudding and @klonos! Thanks @bugfolder for seeing this from start to finish!

**Comment by quicksketch** [comment]
Merged into 1.x and 1.21.x. Thanks!

**Pull request merged**